### PR TITLE
Refactor type substitution solver

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+      if (!method.isConstructor() && substitution.typeVariables().containsAll(method.typeVariableTypes)) {
+        // we found all the substitution
+        break;
+      }
+    }
+    return substitution;
+  }
+
+  private TypeSubstitution inferTypeSubstitution(JavaSymbol.MethodJavaSymbol method, TypeSubstitution currentSubstitution, JavaType formalType, JavaType argumentType,
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java]: MAJOR Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1142)
MAJOR The Cyclomatic Complexity of this method "inferTypeSubstitution" is 24 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    }
+    return currentSubstitution;
+  }
+
+  private static boolean sameErasure(JavaType formalType, JavaType superType) {
+    return formalType.erasure() == superType.erasure();
+  }
+
+  private boolean isNullType(JavaType argType) {
+    return argType == symbols.nullType;
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+      TypeSubstitution newSubstitution = inferTypeSubstitution(method, currentSubstitution, ((WildCardType) formalType).bound, inferedFromArg, variableArity, remainingArgTypes);
+      return mergeTypeSubstitutions(currentSubstitution, newSubstitution);
+    } else {
+      // nothing to infer for simple class types or primitive types
+    }
+    return currentSubstitution;
+  }
+
+  private static boolean sameErasure(JavaType formalType, JavaType superType) {
+    return formalType.erasure() == superType.erasure();
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+  }
+
+  private boolean isNullType(JavaType argType) {
+    return argType == symbols.nullType;
+  }
+
+  private static boolean isRawTypeOfType(JavaType argType, JavaType formalType) {
+    return argType == formalType.erasure();
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* MAJOR 5 major

Watch the comments in this conversation to review them.

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+  private static List<JavaType> listOfTypes(JavaType type, int size) {
+    List<JavaType> result = new ArrayList<>(size);
+    for (int j = 0; j < size; j++) {
+      result.add(type);
+    }
+    return result;
+  }
+
+  private static boolean sameErasure(JavaType type1, JavaType type2) {
+    return type1.erasure() == type2.erasure();
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+          result = mergeTypeSubstitutions(substitution, newSubstitution);
+          break;
+        }
+      }
+    }
+    return result;
+  }
+
+  private static boolean isRawTypeOfType(JavaType rawType, JavaType type) {
+    return rawType == type.erasure();
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+    }
+    return result;
+  }
+
+  private static boolean sameErasure(JavaType type1, JavaType type2) {
+    return type1.erasure() == type2.erasure();
+  }
+
+  private boolean isNullType(JavaType type) {
+    return type == symbols.nullType;
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* MAJOR 3 major

Watch the comments in this conversation to review them.

File: java-frontend/src/main/java/org/sonar/java/resolve/ClassJavaType.java [hunk 8]
```diff
@@ -36,7 +36,11 @@
   List<JavaType> interfaces;
 
   public ClassJavaType(JavaSymbol.TypeJavaSymbol symbol) {
-    super(CLASS, symbol);
+    this(CLASS, symbol);
+  }
+
+  ClassJavaType(int tag, JavaSymbol.TypeJavaSymbol symbol) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/resolve/ClassJavaType.java]: should be protected

File: java-frontend/src/main/java/org/sonar/java/resolve/ParametrizedTypeJavaType.java [hunk 9]
```diff
@@ -31,17 +33,21 @@
...
   }
 
-
   @Override
   public JavaType erasure() {
     return rawType.erasure();
   }
 
+  @Override
+  public boolean isClass() {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/resolve/ParametrizedTypeJavaType.java]: This should probably be tested in java type based on tag for consistency.

**Pull request merged**